catris.py

```diff
@@ -439,127 +521,150 @@ def render_game(self) -> None:
+                display_text += COLOR % 0
+            lines.append(b" " * ((80 - item_width) // 2) + display_text)
+
+        return lines
+
+    def handle_key_press(self, received: bytes) -> bool:
+        i = self._all_menu_items.index(self._selected_item)
+        if received in (UP_ARROW_KEY, b"W", b"w") and i > 0:
+            self._selected_item = self._all_menu_items[i - 1]
+        if received in (DOWN_ARROW_KEY, b"S", b"s"):
+            try:
+                self._selected_item = self._all_menu_items[i + 1]
+            except IndexError:
+                pass
+        if received == b"\r":
+            if self._selected_item == "New Game":
+                assert self._client.name is not None
+                with self._client.server.access_game_state() as state:
+                    player = state.add_player(self._client.name)
+                    self._client.view = PlayingView(self._client, player)
+            elif self._selected_item == "Quit":
+                return True
+            else:
+                raise NotImplementedError(self._selected_item)
 
-            # Send it all at once, so that hopefully cursor won't be in a
-            # temporary place for long times, even if internet is slow
-            to_send = b""
+        return False  # do not quit yet
 
-            for y, (old_line, new_line) in enumerate(
-                zip(self.last_displayed_lines, lines)
-            ):
-                if old_line != new_line:
-                    to_send += MOVE_CURSOR % (y + 1, 1)
-                    to_send += new_line
-                    to_send += CLEAR_TO_END_OF_LINE
-            self.last_displayed_lines = lines.copy()
 
-            # Wipe bottom of terminal and leave cursor there.
-            # This way, if user types something, it will be wiped.
-            to_send += MOVE_CURSOR % (24, 1)
-            to_send += CLEAR_TO_END_OF_LINE
+class Client(socketserver.BaseRequestHandler):
+    server: Server
+    request: socket.socket
+
+    def setup(self) -> None:
+        print(self.client_address, "New connection")
+        self.last_displayed_lines: list[bytes] = []
+        self.send_queue: queue.Queue[bytes | None] = queue.Queue()
+        self.name: str | None = None
+        self.view: AskNameView | PlayingView | GameOverView = AskNameView(self)
+
+    def render(self) -> None:
+        # Bottom of terminal. If user types something, it's unlikely to be
+        # noticed here before it gets wiped by the next refresh.
+        cursor_pos = (24, 1)
+
+        if isinstance(self.view, AskNameView):
+            lines, cursor_pos = self.view.get_lines_to_render_and_cursor_pos()
+        else:
+            lines = self.view.get_lines_to_render()
+
+        while len(lines) < len(self.last_displayed_lines):
+            lines.append(b"")
+        while len(lines) > len(self.last_displayed_lines):
+            self.last_displayed_lines.append(b"")
```

Comment:
Maybe instead make 24 displayed lines in `setup()` and assert there's less than that. More is a bug anyway, because I aim to support 80x24 terminals

Reply:
This could create subtle bugs in the future that error instead of just not showing everything. Seems like a bad idea.